x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java

```diff
@@ -348,6 +352,63 @@ public void authorizeIndexAction(RequestInfo requestInfo, AuthorizationInfo auth
         }
     }
 
+    private boolean isChildActionAuthorizedByParent(RequestInfo requestInfo, AuthorizationInfo authorizationInfo) {
+        final AuthorizationContext parent = requestInfo.getParentAuthorizationContext();
+        if (parent == null) {
+            return false;
+        }
+
+        final AuthorizationResult parentResult = parent.getAuthorizationResult();
+        if ((parentResult instanceof IndexAuthorizationResult) == false) {
+            return false;
+        }
+        final IndicesAccessControl indicesAccessControl = ((IndexAuthorizationResult) parentResult).getIndicesAccessControl();
+        if (indicesAccessControl == null) {
+            // This can happen for is the parent request was authorized by index name only - e.g. bulk request
+            // A missing IAC is not an error, but it means we can't safely tie authz of the child action to the parent authz
+            return false;
+        }
+
+        if (requestInfo.getAction().startsWith(parent.getAction()) == false) {
+            // Parent action is not a true parent
+            // We want to treat shard level actions (those that append '[s]' and/or '[p]' & '[r]')
+            // or similar (e.g. search phases) as children, but not every action that is triggered
+            // within another action should be authorized this way
+            return false;
+        }
+
+        if (authorizationInfo.equals(parent.getAuthorizationInfo()) == false) {
+            // Authorization changed
+            // This should only happen if the user's list of roles changed between requests
+            // Take the safe option and perform full authorization
+            return false;
+        }
+
+        final IndicesRequest indicesRequest;
+        if (requestInfo.getRequest() instanceof IndicesRequest) {
+            indicesRequest = (IndicesRequest) requestInfo.getRequest();
+        } else {
+            // Can only handle indices request here
+            return false;
+        }
+
+        final String[] indices = indicesRequest.indices();
+        if (indices == null || indices.length == 0) {
+            // No indices to check
+            return false;
+        }
+
+        return Arrays.stream(indices).allMatch(idx -> {
```

Comment:
nit: Just because we want to squeeze out performance I would use a classic for loop here.